File: src/main/java/net/hectus/neobb/lore/PersonItemLoreBuilder.java
package net.hectus.neobb.lore;

import com.marcpg.libpg.lang.Translation;
import net.hectus.neobb.turn.default_game.attributes.function.BuffFunction;
import net.hectus.neobb.turn.default_game.attributes.function.CounterFunction;
import net.hectus.neobb.turn.person_game.categorization.Category;
import net.hectus.neobb.turn.person_game.warp.PWarpTurn;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.TextDecoration;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class PersonItemLoreBuilder extends ItemLoreBuilder {
    @Override
    public List<Component> build(Locale l) {
        if (!(turn instanceof Category category)) return List.of();

        List<Component> lore = new ArrayList<>();

        if (turn.damage() > 0.0) {
            lore.add(Component.text("Damage: ", category.categoryColor(), TextDecoration.BOLD)
                    .append(Component.text(turn.damage() + "hearts.")));
        }

        if (turn instanceof CounterFunction counter) {
            lore.add(Translation.component(l, "item-lore.counters").color(category.categoryColor()).decorate(TextDecoration.BOLD)
                    .append(Component.text(String.join(", ", counter.counters().stream().map(c -> c.text(l)).toList()))));
        }
        if (turn instanceof PWarpTurn warp) {
            lore.add(Component.text("Probability: ", category.categoryColor(), TextDecoration.BOLD)
                    .append(Component.text(warp.chance() + "%")));
        }
        if (turn instanceof BuffFunction buff) {
            lore.addAll(buff.buffs().stream().map(b -> Component.text("Probability: ", category.categoryColor(), TextDecoration.BOLD)
                    .append(Component.text(b.text(l) + b.target(l), b.color()))).toList());
        }

        return lore;
    }
}

File: src/main/kotlin/net/hectus/neobb/lore/PersonItemLoreBuilder.kt
package net.hectus.neobb.lore

import net.hectus.neobb.turn.default_game.attribute.function.BuffFunction
import net.hectus.neobb.turn.default_game.attribute.function.CounterFunction
import net.hectus.neobb.turn.person_game.categorization.Category
import net.hectus.neobb.turn.person_game.warp.PWarpTurn
import net.hectus.neobb.util.component
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.TextDecoration
import java.util.*

class PersonItemLoreBuilder: ItemLoreBuilder() {
    override fun build(locale: Locale): List<Component> {
        if (turn !is Category) return emptyList()

        val lore = mutableListOf<Component>()

        if ((turn?.damage ?: 0.0) > 0.0) {
            lore.add(
                Component.text("Damage: ", (turn as Category).categoryColor, TextDecoration.BOLD)
                    .append(Component.text("${turn!!.damage} hearts."))
            )
        }

        if (turn is CounterFunction) {
            lore.add(locale.component("item-lore.counters", color = (turn as Category).categoryColor, decoration = TextDecoration.BOLD)
                .append(Component.text((turn as CounterFunction).counters().joinToString(", ") { it.text(locale) })))
        }
        if (turn is PWarpTurn) {
            lore.add(
                Component.text("Probability: ", (turn as Category).categoryColor, TextDecoration.BOLD)
                    .append(Component.text((turn as PWarpTurn).chance.toString() + "%"))
            )
        }
        if (turn is BuffFunction) {
            lore.addAll((turn as BuffFunction).buffs().map { b -> Component.text("Probability: ", (turn as Category).categoryColor, TextDecoration.BOLD)
                .append(Component.text(b.text(locale) + b.target(locale), b.color())) }.toList())
        }

        return lore
    }
}
